mllib/src/main/scala/org/apache/spark/mllib/clustering/metrics/FastEuclideanOps.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.clustering.metrics
+
+import breeze.linalg.{ DenseVector => BDV, SparseVector => BSV, Vector => BV }
+
+import org.apache.spark.mllib.base._
+import org.apache.spark.mllib.linalg.{ SparseVector, DenseVector, Vector }
+import org.apache.spark.mllib.base.{ Centroid, FPoint, PointOps, Infinity, Zero }
+
+class FastEUPoint(raw: BV[Double], weight: Double) extends FPoint(raw, weight) {
+  val norm = if (weight == Zero) Zero else raw.dot(raw) / (weight * weight)
```

Comment:
It is most efficient to maintain the cluster centers in homogeneous
coordinates.

On Thu, Jan 8, 2015 at 3:56 PM, Xiangrui Meng <notifications@github.com>
wrote:

> In
> mllib/src/main/scala/org/apache/spark/mllib/clustering/metrics/FastEuclideanOps.scala
> <https://github.com/apache/spark/pull/2634#discussion-diff-22693112>:
>
> > + * distributed under the License is distributed on an "AS IS" BASIS,
> > + * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > + * See the License for the specific language governing permissions and
> > + * limitations under the License.
> > + */
> > +
> > +package org.apache.spark.mllib.clustering.metrics
> > +
> > +import breeze.linalg.{ DenseVector => BDV, SparseVector => BSV, Vector => BV }
> > +
> > +import org.apache.spark.mllib.base._
> > +import org.apache.spark.mllib.linalg.{ SparseVector, DenseVector, Vector }
> > +import org.apache.spark.mllib.base.{ Centroid, FPoint, PointOps, Infinity, Zero }
> > +
> > +class FastEUPoint(raw: BV[Double], weight: Double) extends FPoint(raw, weight) {
> > +  val norm = if (weight == Zero) Zero else raw.dot(raw) / (weight * weight)
>
> Should weight be only used in aggregation rather than distance
> computation?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/2634/files#r22693112>.
>